src/main.py

```diff
@@ -185,49 +188,79 @@ def main(args):
 
     with strategy.scope():
         print("######### Preparing model #########")
-        model = Wav2Vec2ForCTCTrainer.from_pretrained(
-            args.model_id,
-            jit_compile=jit_compile,
-            input_shape=model_input_shape,
-            apply_spec_augment=args.apply_spec_augment,
-            survival_prob=args.survival_prob,
-        )
-
-        # during fine-tuning, we need to freeze the feature extraction layer from the pre-trained weights
-        model.freeze_feature_extractor()
+        model = build_model(args)
 
         # `division_factor` should be passed else loss will be summed
         # it will help us in distributed training over several processes
-        loss_fn = CTCLoss(
+        loss = CTCLoss(
             model.config, model_input_shape, division_factor=global_batch_size
         )
 
-        optimizer = tf.keras.optimizers.Adam(learning_rate=args.lr)
-        model.compile(optimizer, loss_fn)
-
-    model.summary()
-    print("######### Initiating training #########")
-    model.fit(
-        tr_dataset,
-        validation_data=val_dataset,
-        epochs=args.max_epochs,
-        callbacks=fetch_callbacks(args),
-        verbose="auto",
-    )
-
-    print("\n######### Preparing for evaluation #########")
+        ######################### STAGE-1 #########################
+
+        print("######## FREEZING ########")
+        # till `stage1_epochs`, we will train only `lm_head`
+        model.layers[0].trainable = False
+        print(model.layers[0])
+        print("#########################")
+        model.summary()
+
+        optimizer = tf.keras.optimizers.Adam(learning_rate=args.stage1_lr)
+        model.compile(optimizer=optimizer, loss=loss)
+
+        try:
+            history = model.fit(
+                tr_dataset,
+                validation_data=val_dataset,
+                epochs=args.stage1_epochs,
+                callbacks=fetch_callbacks(args, is_stage2=False),
+                verbose="auto",
+            )
+            print(history.history)
+        except KeyboardInterrupt:
+            print("Interrupting through KEYBOARD")
+
+        ###########################################################
+
+        ######################### STAGE-2 #########################
+        # during fine-tuning, it's recommended to freeze the feature extraction layer from the pre-trained weights
+
+        model.trainable = True
+        print("############## FREEZING ##############")
+        for i in range(len(model.layers[0].layers) - 2):
+            model.layers[0].layers[i].trainable = False
+            print(model.layers[0].layers[i])
```

Comment:
I am a bit confused about this. During fine-tuning, we actually unfreeze the middle layers, pre-final layers and then continue the process with smaller learning rates so that the pre-trained weights do not break too fast. But what's the notion here? 

Also, since you're already doing `model.layers[0].trainable = False` then by default, all the layers under it are getting non-trainable.  Know more about this here: https://keras.io/guides/transfer_learning/.

Reply:
In speech models like Wav2Vec2, it's recommened to keep feature extraction layers (initial convolutional layers) freezed throughout the fine-tuning process. For fine-tuning also, we just need to train all transformer blocks & LM layer (top most layer).

Here training is divided into 2 stages:
1. freeze the pre-trained model and only train the LM head until model converges
2. unfreeze complete model (keep conv feature extraction layer freezed) and train the model

`model.layers[0].trainable = False` is helping in first stage but for 2nd stage we are unfreezing complete model again using `model.trainable = True`.

Reply:
Clears my confusion. We might want to add extensive comments in this part.